File: app/src/main/java/com/komica/reader/util/FavoritesManager.java
package com.komica.reader.util;

import android.content.Context;
import android.content.SharedPreferences;
import java.util.HashSet;
import java.util.Set;
import java.util.Collections;

public class FavoritesManager {
    private static final String PREF_NAME = "KomicaFavorites";
    private static final String KEY_FAVORITES = "favorite_boards";
    private static FavoritesManager instance;
    private final SharedPreferences sharedPreferences;
    
    // Memory cache to avoid repeated disk reads
    private volatile Set<String> cachedFavorites;

    private FavoritesManager(Context context) {
        sharedPreferences = context.getApplicationContext().getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        // Initial load
        cachedFavorites = new HashSet<>(sharedPreferences.getStringSet(KEY_FAVORITES, new HashSet<>()));
    }

    public static synchronized FavoritesManager getInstance(Context context) {
        if (instance == null) {
            instance = new FavoritesManager(context);
        }
        return instance;
    }

    public synchronized Set<String> getFavorites() {
        return Collections.unmodifiableSet(cachedFavorites);
    }

    public boolean isFavorite(String boardUrl) {
        return cachedFavorites.contains(boardUrl);
    }

    public synchronized void toggleFavorite(String boardUrl) {
        Set<String> newFavorites = new HashSet<>(cachedFavorites);
        if (newFavorites.contains(boardUrl)) {
            newFavorites.remove(boardUrl);
        } else {
            newFavorites.add(boardUrl);
        }
        
        cachedFavorites = newFavorites;
        sharedPreferences.edit().putStringSet(KEY_FAVORITES, newFavorites).apply();
    }
}

File: app/src/main/java/com/komica/reader/util/FavoritesManager.kt
package com.komica.reader.util

import android.content.Context
import android.content.SharedPreferences
import java.util.Collections
import java.util.HashSet

// 繁體中文註解：我的最愛管理器，使用快取減少磁碟讀取
class FavoritesManager private constructor(context: Context) {
    private val sharedPreferences: SharedPreferences =
        context.applicationContext.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)

    @Volatile
    private var cachedFavorites: Set<String> =
        HashSet(sharedPreferences.getStringSet(KEY_FAVORITES, emptySet()) ?: emptySet())

    fun getFavorites(): Set<String> {
        return Collections.unmodifiableSet(cachedFavorites)
    }

    fun isFavorite(boardUrl: String): Boolean {
        return cachedFavorites.contains(boardUrl)
    }

    @Synchronized
    fun replaceFavorites(newFavorites: Set<String>) {
        cachedFavorites = HashSet(newFavorites)
        // 繁體中文註解：完整覆蓋最愛清單，確保還原內容一致
        sharedPreferences.edit().putStringSet(KEY_FAVORITES, cachedFavorites).commit()
    }

    @Synchronized
    fun toggleFavorite(boardUrl: String) {
        val newFavorites = HashSet(cachedFavorites)
        if (newFavorites.contains(boardUrl)) {
            newFavorites.remove(boardUrl)
        } else {
            newFavorites.add(boardUrl)
        }
        cachedFavorites = newFavorites
        // 繁體中文註解：資料量小，使用 commit 確保立即寫入
        sharedPreferences.edit().putStringSet(KEY_FAVORITES, newFavorites).commit()
    }

    companion object {
        private const val PREF_NAME = "KomicaFavorites"
        private const val KEY_FAVORITES = "favorite_boards"

        @Volatile
        private var instance: FavoritesManager? = null

        @JvmStatic
        fun getInstance(context: Context): FavoritesManager {
            return instance ?: synchronized(this) {
                instance ?: FavoritesManager(context).also { instance = it }
            }
        }
    }
}
